File: 프로그래머스/연습문제/Lv2/땅따먹기/땅따먹기.py
def solution(land):
    DP = [[0 for i in range(len(land[0]))] for i in range(len(land))]
    for i in range(len(land[0])):
        DP[0][i] = land[0][i]

    for i in range(1, len(land)):
        for j in range(len(land[0])):
            max_n = 0
            for k in range(len(land[0])):
                if j == k: continue
                else: max_n = max(DP[i-1][k], max_n)
            DP[i][j] = max_n+land[i][j]

    return max(DP[-1])

File: 프로그래머스/연습문제/Lv2/땅따먹기/땅따먹기.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int solution(vector<vector<int> > land){
    vector<vector<int>> DP;
    DP.push_back(land[0]);
    for(int i=1; i<land.size(); i++){
        vector<int> v(land[0].size(), 0);
        DP.push_back(v);
    }
    
    for(int i=1; i<land.size(); i++){
        for(int j=0; j<land[0].size(); j++){
            int max = 0;
            for(int k=0; k<land[0].size(); k++){
                if(j == k) continue;
                else if(DP[i-1][k] > max) max = DP[i-1][k];
            }
            DP[i][j] = max+land[i][j];
        }
    }
    
    return *max_element(DP[land.size()-1].begin(), DP[land.size()-1].end());
}
